File: ios/FastingTimerPro/Sources/FastingProgress.swift
import Foundation

struct FastingProgress {
    let elapsed: TimeInterval
    let remaining: TimeInterval
    let progress01: Double
    let currentStage: FastingStage
    let nextStage: FastingStage?
}

enum FastingProgressCalculator {

    static func compute(
        session: FastingSession,
        now: Date,
        stages: [FastingStage] = FastingStages.default()
    ) -> FastingProgress {
        let elapsed: TimeInterval = Swift.max(0.0, now.timeIntervalSince(session.startTime))
        let total: TimeInterval = Swift.max(60.0, session.plannedDuration)
        let remaining: TimeInterval = Swift.max(0.0, session.endTime.timeIntervalSince(now))
        let progress01: Double = Swift.min(1.0, Swift.max(0.0, elapsed / total))

        let elapsedHours = elapsed / 3600.0
        var current = stages.first!
        for stage in stages {
            if elapsedHours >= stage.startHours {
                if let end = stage.endHours {
                    if elapsedHours < end {
                        current = stage
                    }
                } else {
                    current = stage
                }
            }
        }

        let next = stages.first(where: { $0.startHours > current.startHours })

        return FastingProgress(
            elapsed: elapsed,
            remaining: remaining,
            progress01: progress01,
            currentStage: current,
            nextStage: next
        )
    }
}

File: android/FastingTimerPro/app/src/main/java/com/fastingtimerpro/app/domain/FastingProgress.kt
package com.fastingtimerpro.app.domain

import java.time.Duration
import java.time.Instant

data class FastingProgress(
    val elapsed: Duration,
    val remaining: Duration,
    val progress01: Float,
    val currentStage: FastingStage,
    val nextStage: FastingStage?
)

object FastingProgressCalculator {

    fun compute(
        session: FastingSession,
        now: Instant,
        stages: List<FastingStage> = FastingStages.defaultStages()
    ): FastingProgress {
        val rawElapsed = Duration.between(session.startTime, now)
        val elapsed = if (rawElapsed.isNegative) Duration.ZERO else rawElapsed

        val total = if (session.plannedDuration < Duration.ofMinutes(1)) {
            Duration.ofMinutes(1)
        } else {
            session.plannedDuration
        }

        val rawRemaining = Duration.between(now, session.endTime)
        val remaining = if (rawRemaining.isNegative) Duration.ZERO else rawRemaining
        val progress01 = (elapsed.toMillis().toDouble() / total.toMillis().toDouble())
            .coerceIn(0.0, 1.0)
            .toFloat()

        val elapsedHours = elapsed.toMinutes().toDouble() / 60.0
        val current = stages.lastOrNull { stage ->
            elapsedHours >= stage.startHours && (stage.endHours == null || elapsedHours < stage.endHours)
        } ?: stages.first()

        val next = stages.firstOrNull { stage ->
            stage.startHours > current.startHours
        }

        return FastingProgress(
            elapsed = elapsed,
            remaining = remaining,
            progress01 = progress01,
            currentStage = current,
            nextStage = next
        )
    }
}
